demo/src/components/canvas3d.js

```diff
@@ -0,0 +1,211 @@
+  const [x, y, z] = vector;
+  return [
+    x,
+    y * cosA - z * sinA,
+    y * sinA + z * cosA
+  ];
+}
+
+// rotate vector around y axis
+// @param {number} angle: in degrees
+function rotateY(vector, angle) {
+  const r = angle / 180 * Math.PI;
+  const cosA = Math.cos(r);
+  const sinA = Math.sin(r);
+  const [x, y, z] = vector;
+  return [
+    x * cosA + z * sinA,
+    y,
+    -x * sinA + z * cosA
+  ];
+}
+
+const ua = typeof window.navigator !== 'undefined' ?
+  window.navigator.userAgent.toLowerCase() : '';
+const firefox = ua.indexOf('firefox') !== -1;
+
+/* Interaction */
+
+export default class Canvas3D extends Component {
+
+  static getViewport({width, height, lookAt, distance, rotationX, rotationY, fov}) {
+    let cameraDir = [0, 0, distance];
+    cameraDir = rotateX(cameraDir, rotationX);
+    cameraDir = rotateY(cameraDir, rotationY);
+    return new PerspectiveViewport({
+      width,
+      height,
+      lookAt,
+      fovy: fov,
+      eye: add(lookAt, cameraDir)
+    });
+  }
+
+  constructor(props) {
+    super(props);
+    this._dragStartPos = null;
+  }
+
+  @autobind _onDragStart(evt) {
+    const {pageX, pageY} = evt;
+    this._dragStartPos = [pageX, pageY];
+    this.props.onViewportChange({isDragging: true});
+  }
+
+  @autobind _onDrag(evt) {
+    if (this._dragStartPos) {
+      const {pageX, pageY} = evt;
+      const {width, height} = this.props;
+      const dx = (pageX - this._dragStartPos[0]) / width;
+      const dy = (pageY - this._dragStartPos[1]) / height;
+
+      if (evt.shiftKey || evt.ctrlKey || evt.altKey || evt.metaKey) {
+        // rotate
+        const {rotationX, rotationY} = this.props;
+        const newRotationX = clamp(rotationX - dy * 180, -90, 90);
+        const newRotationY = (rotationY - dx * 180) % 360;
+
+        this.props.onViewportChange({
+          rotationX: newRotationX,
+          rotationY: newRotationY
+        });
+      } else {
+        // pan
+        const {lookAt, distance, rotationX, rotationY, fov} = this.props;
+
+        const unitsPerPixel = distance / Math.tan(fov / 180 * Math.PI / 2) / 2;
+        let translation = [-unitsPerPixel * dx, unitsPerPixel * dy, 0];
+        translation = rotateX(translation, rotationX);
+        translation = rotateY(translation, rotationY);
+        this.props.onViewportChange({
+          lookAt: add(lookAt, translation)
+        });
+      }
+
+      this._dragStartPos = [pageX, pageY];
+    }
+  }
+
+  @autobind _onDragEnd() {
+    this._dragStartPos = null;
+    this.props.onViewportChange({isDragging: false});
+  }
+
+  @autobind _onWheel(evt) {
+    evt.preventDefault();
+    let value = evt.deltaY;
+    // Firefox doubles the values on retina screens...
+    if (firefox && evt.deltaMode === window.WheelEvent.DOM_DELTA_PIXEL) {
+      value /= window.devicePixelRatio;
+    }
+    if (evt.deltaMode === window.WheelEvent.DOM_DELTA_LINE) {
+      value *= 40;
+    }
+    if (value !== 0 && value % 4.000244140625 === 0) {
```

Comment:
This originated from react-map-gl. Chrome on desktop produces wheel deltas like this, and it's (apparently) an effective way to differentiate between mouse and trackpad input.